eitprocessing/features/breath_detection.py

```diff
@@ -0,0 +1,471 @@
+            label=result_label,
+            name="Breaths as determined by BreathDetection",
+            unit=None,
+            category="breath",
+            intervals=[(breath.start_time, breath.end_time) for breath in breaths],
+            values=breaths,
+            parameters={type(self): dict(vars(self))},
+            derived_from=[continuous_data],
+        )
+
+        if store:
+            sequence.interval_data.add(breaths_container)
+
+        return breaths_container
+
+    def _detect_invalid_data(self, data: np.ndarray) -> np.ndarray:
+        """Detects invalid data as outliers outside an upper and lower cutoff.
+
+        This function defines a lower and upper cutoff. Data beyond those cutoffs is considered invalid for the purposes
+        of breath detection.
+
+        The lower cutoff is a distance away from the mean. The distance is m times the distance between the mean and the
+        nth percentile of the data. The upper cutoff is m times the distance between the mean and the (100 - n)th
+        percentile. m is given by `invalid_data_removal_multiplier` and n is given by `invalid_data_removal_percentile`.
+
+        For example, with m = 4 and n = 5, the mean = 100, 5% of the data is below/equal to 90, and 5% of the data is
+        above/equal to 120, all data below 100 - (4 * 10) = 60 and above 100 + (4 * 20) = 180 is considerd invalid.
+
+        Args:
+            data (np.ndarray): _description_
+
+        Returns:
+            np.ndarray: the indices of the data points with values outside the lower and upper cutoff values.
+        """
+        data_mean = np.mean(data)
+
+        lower_percentile = np.percentile(data, self.invalid_data_removal_percentile)
+        cutoff_low = data_mean - (data_mean - lower_percentile) * self.invalid_data_removal_multiplier
+
+        upper_percentile = np.percentile(data, 100 - self.invalid_data_removal_percentile)
+        cutoff_high = data_mean + (upper_percentile - data_mean) * self.invalid_data_removal_multiplier
+
+        # detect indices of outliers
+        return np.flatnonzero((data < cutoff_low) | (data > cutoff_high))
+
+    def _remove_invalid_data(self, data: np.ndarray, invalid_data_indices: np.ndarray) -> np.ndarray:
+        """Removes invalid data points and replace them with the nearest non-np.nan value.."""
+        data = np.copy(data)
+        data[invalid_data_indices] = np.nan
+        return self._fill_nan_with_nearest_neighbour(data)
+
+    def _detect_peaks_and_valleys(self, data: np.ndarray) -> tuple[np.ndarray, np.ndarray]:
+        window_size = int(self.sample_frequency * self.averaging_window_duration)
+        averager = MovingAverage(window_size=window_size, window_function=self.averaging_window_function)
+        moving_average = averager.apply(data)
+
+        peak_indices = self._find_features(data, moving_average)
+        valley_indices = self._find_features(data, moving_average, invert=True)
```

Comment:
Wouldn't it be more concise to make this function return both in one go? It's also a bit confusing that it's called "features" if it's specifically the extremes, and `invert` further implies that it's specifically maxima.

```suggestion
        peak_indices, valley_indices = self._find_extremes(data, moving_average)
```


Given that `peak_indices, valley_indices` are passed together throughout, they could even be given/passed as a single variable:
```suggestion
        extremes = self._find_extremes(data, moving_average)
```

Reply:
I don't want it to return both at the same time. The whole point here is to have the method do one thing, and that is detect peaks. The function `_detect_peaks_and_valleys()` ties it together and does both. 

I agree the name _feature_ is too vague here. The name _extremes_ for me has the connotation of an outlier, but _extrema_ maybe more specifically indicates local features.